In [1]:
#import the libraries
import re
import pandas as pd
import latex

# Correlation coefficient

The correlation coefficient measures the strength of the linear relationship betwwn two quantitative variables.
The correlation coefficient is usually denoted as $r$

To compute the correlation $r$ between $x$ and $y$, this is the formula

$$r = \frac 1 {n - 1} \sum_{i=1}^n \left(\frac{x_i - {\bar x}} {s_x} \right) \left(\frac{y_i - {\bar y}} {s_y}\right)$$ 


$$r = \frac 1 {n - 1} \left[\sum_{i=1}^n(x_i - {\bar x}) (y_i - {\bar y})\right] / {s_x s_y}$$

In [11]:
df = pd.DataFrame({'year': [1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003],
                  'Number of Casino Employees $x$ (thousands)': [20,23,29,27,30,34,35,37,40,43],
                  'Crime Rate $y$ (number of crimes per 1.000 population)': [1.32,1.67,2.17,2.70,2.75,2.87,3.65,2.86,3.61,4.25]})
df

year  Number of Casino Employees $x$ (thousands)  \
0  1994                                          20   
1  1995                                          23   
2  1996                                          29   
3  1997                                          27   
4  1998                                          30   
5  1999                                          34   
6  2000                                          35   
7  2001                                          37   
8  2002                                          40   
9  2003                                          43   

   Crime Rate $y$ (number of crimes per 1.000 population)  
0                                               1.32       
1                                               1.67       
2                                               2.17       
3                                               2.70       
4                                               2.75       
5                                               2.87       
6                                               3.65       
7                                               2.86       
8                                               3.61       
9                                               4.25

## Assignment:

compute the correlation coefficient for the data in df

## Solution: 

First let's make the formula in a way that we can calculate this easily. 

$r = \frac {\sum_{i=1}^n(x_i-{\bar x}) (y_i - {\bar y})} {\root \of{\sum_{i=1}^n(x_i-{\bar x})^2 \sum_{i=1}^n(y_i-{\bar y})^2}}$

In [12]:
#Right, now let's add different columns to this df, with the equations of the formula calculated in it
#let's start with calculating the mean for x and y

df.describe()

year  Number of Casino Employees $x$ (thousands)  \
count    10.00000                                   10.000000   
mean   1998.50000                                   31.800000   
std       3.02765                                    7.345445   
min    1994.00000                                   20.000000   
25%    1996.25000                                   27.500000   
50%    1998.50000                                   32.000000   
75%    2000.75000                                   36.500000   
max    2003.00000                                   43.000000   

       Crime Rate $y$ (number of crimes per 1.000 population)  
count                                          10.000000       
mean                                            2.785000       
std                                             0.904559       
min                                             1.320000       
25%                                             2.302500       
50%                                             2.805000       
75%                                             3.425000       
max                                             4.250000

In [13]:
#so the mean for x is 31.8 and for y is 2.785
#Now let's add a column with x - the mean of x (31.8)

values = []
for x in df['Number of Casino Employees $x$ (thousands)']:
    values.append(x-31.8)
df.insert(loc = 3, column = '$$x-{\bar_x}$$', value = values)
df

year  Number of Casino Employees $x$ (thousands)  \
0  1994                                          20   
1  1995                                          23   
2  1996                                          29   
3  1997                                          27   
4  1998                                          30   
5  1999                                          34   
6  2000                                          35   
7  2001                                          37   
8  2002                                          40   
9  2003                                          43   

   Crime Rate $y$ (number of crimes per 1.000 population)  $$x-{ar_x}$$  
0                                               1.32               -11.8  
1                                               1.67                -8.8  
2                                               2.17                -2.8  
3                                               2.70                -4.8  
4                                               2.75                -1.8  
5                                               2.87                 2.2  
6                                               3.65                 3.2  
7                                               2.86                 5.2  
8                                               3.61                 8.2  
9                                               4.25                11.2

In [14]:
#Now let's do the same for y (y - 2.785)

values = []
for y in df['Crime Rate $y$ (number of crimes per 1.000 population)']:
    values.append(y-2.785)
df.insert(loc = 4, column = '$$y-{\bar_y}$$', value = values)
df

year  Number of Casino Employees $x$ (thousands)  \
0  1994                                          20   
1  1995                                          23   
2  1996                                          29   
3  1997                                          27   
4  1998                                          30   
5  1999                                          34   
6  2000                                          35   
7  2001                                          37   
8  2002                                          40   
9  2003                                          43   

   Crime Rate $y$ (number of crimes per 1.000 population)  $$x-{ar_x}$$  \
0                                               1.32               -11.8   
1                                               1.67                -8.8   
2                                               2.17                -2.8   
3                                               2.70                -4.8   
4                                               2.75                -1.8   
5                                               2.87                 2.2   
6                                               3.65                 3.2   
7                                               2.86                 5.2   
8                                               3.61                 8.2   
9                                               4.25                11.2   

   $$y-{ar_y}$$  
0         -1.465  
1         -1.115  
2         -0.615  
3         -0.085  
4         -0.035  
5          0.085  
6          0.865  
7          0.075  
8          0.825  
9          1.465

In [16]:
#Great, now this is done, so we can add a column with the last two columns multiplied:

df['x*y'] = df['$$x-{\bar_x}$$'] * df['$$y-{\bar_y}$$']
df

year  Number of Casino Employees $x$ (thousands)  \
0  1994                                          20   
1  1995                                          23   
2  1996                                          29   
3  1997                                          27   
4  1998                                          30   
5  1999                                          34   
6  2000                                          35   
7  2001                                          37   
8  2002                                          40   
9  2003                                          43   

   Crime Rate $y$ (number of crimes per 1.000 population)  $$x-{ar_x}$$  \
0                                               1.32               -11.8   
1                                               1.67                -8.8   
2                                               2.17                -2.8   
3                                               2.70                -4.8   
4                                               2.75                -1.8   
5                                               2.87                 2.2   
6                                               3.65                 3.2   
7                                               2.86                 5.2   
8                                               3.61                 8.2   
9                                               4.25                11.2   

   $$y-{ar_y}$$     x*y  
0         -1.465  17.287  
1         -1.115   9.812  
2         -0.615   1.722  
3         -0.085   0.408  
4         -0.035   0.063  
5          0.085   0.187  
6          0.865   2.768  
7          0.075   0.390  
8          0.825   6.765  
9          1.465  16.408

In [17]:
#Super, now let's add two more columns with the differences of x squared and the same for y:

df['$x^2$'] = df['$$x-{\bar_x}$$']**2
df['$y^2$'] = df['$$y-{\bar_y}$$']**2
df

year  Number of Casino Employees $x$ (thousands)  \
0  1994                                          20   
1  1995                                          23   
2  1996                                          29   
3  1997                                          27   
4  1998                                          30   
5  1999                                          34   
6  2000                                          35   
7  2001                                          37   
8  2002                                          40   
9  2003                                          43   

   Crime Rate $y$ (number of crimes per 1.000 population)  $$x-{ar_x}$$  \
0                                               1.32               -11.8   
1                                               1.67                -8.8   
2                                               2.17                -2.8   
3                                               2.70                -4.8   
4                                               2.75                -1.8   
5                                               2.87                 2.2   
6                                               3.65                 3.2   
7                                               2.86                 5.2   
8                                               3.61                 8.2   
9                                               4.25                11.2   

   $$y-{ar_y}$$     x*y   $x^2$     $y^2$  
0         -1.465  17.287  139.24  2.146225  
1         -1.115   9.812   77.44  1.243225  
2         -0.615   1.722    7.84  0.378225  
3         -0.085   0.408   23.04  0.007225  
4         -0.035   0.063    3.24  0.001225  
5          0.085   0.187    4.84  0.007225  
6          0.865   2.768   10.24  0.748225  
7          0.075   0.390   27.04  0.005625  
8          0.825   6.765   67.24  0.680625  
9          1.465  16.408  125.44  2.146225

In [18]:
#Cool, we are almost there, let's add a total (=sum) row!

df.loc['Total']= df.sum()
df

year  Number of Casino Employees $x$ (thousands)  \
0       1994.0                                        20.0   
1       1995.0                                        23.0   
2       1996.0                                        29.0   
3       1997.0                                        27.0   
4       1998.0                                        30.0   
5       1999.0                                        34.0   
6       2000.0                                        35.0   
7       2001.0                                        37.0   
8       2002.0                                        40.0   
9       2003.0                                        43.0   
Total  19985.0                                       318.0   

       Crime Rate $y$ (number of crimes per 1.000 population)  $$x-{ar_x}$$  \
0                                                   1.32       -1.180000e+01   
1                                                   1.67       -8.800000e+00   
2                                                   2.17       -2.800000e+00   
3                                                   2.70       -4.800000e+00   
4                                                   2.75       -1.800000e+00   
5                                                   2.87        2.200000e+00   
6                                                   3.65        3.200000e+00   
7                                                   2.86        5.200000e+00   
8                                                   3.61        8.200000e+00   
9                                                   4.25        1.120000e+01   
Total                                              27.85       -7.105427e-15   

       $$y-{ar_y}$$     x*y   $x^2$     $y^2$  
0      -1.465000e+00  17.287  139.24  2.146225  
1      -1.115000e+00   9.812   77.44  1.243225  
2      -6.150000e-01   1.722    7.84  0.378225  
3      -8.500000e-02   0.408   23.04  0.007225  
4      -3.500000e-02   0.063    3.24  0.001225  
5       8.500000e-02   0.187    4.84  0.007225  
6       8.650000e-01   2.768   10.24  0.748225  
7       7.500000e-02   0.390   27.04  0.005625  
8       8.250000e-01   6.765   67.24  0.680625  
9       1.465000e+00  16.408  125.44  2.146225  
Total  -1.332268e-15  55.810  485.60  7.364050

Now let's use these outcomes to fill in the formula: 

- $r = \frac {\sum_{i=1}^n(x_i-{\bar x}) (y_i - {\bar y})} {\root \of{\sum_{i=1}^n(x_i-{\bar x})^2 \sum_{i=1}^n(y_i-{\bar y})^2}}$
- $r = \frac {55.810} {\root \of{({485.6})({7.3641})}}$
- $r = {0.933}$

Conclusion, the correlation is positive, but this does not mean that an increase in the number of casino workers **causes** and increase in the crime rate. 